pkg/ipamd/datastore/data_store.go

```diff
@@ -841,6 +1112,39 @@ func (ds *DataStore) FreeableIPs(eniID string) []string {
 	return freeable
 }
 
+// FreeablePrefixes returns a list of unused and potentially freeable IPs.
+// Note result may already be stale by the time you look at it.
+func (ds *DataStore) FreeablePrefixes(eniID string) []string {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	eni := ds.eniPool[eniID]
+	if eni == nil {
+		// Can't free any IPs from an ENI we don't know about...
+		return []string{}
+	}
+	ds.log.Infof("In freeable prefix for eni %s", eniID)
+
+	freeable := make([]string, 0, len(eni.IPv4Prefixes))
+	for _, prefix := range eni.IPv4Prefixes {
+		DBlen := (int)(prefix.AllocatedIPs.IPsPerPrefix / 8)
+
+		ds.log.Infof("In get IP from prefix - %s and octet len %d", prefix, DBlen)
+		var octet int
+		for octet = 0; octet < DBlen; octet++ {
+			if prefix.AllocatedIPs.UsedIPs[octet]&0xFF != 0 {
+				break
+			}
+		}
```

Comment:
This should be a method on UsedIPs (or AllocatedIPs) - don't leak the (slightly complicated / error-prone) representation outside those classes.

Reply:
Yeah agreed.